contributing/specifications/functions.md

```diff
@@ -0,0 +1,111 @@
+
+A Function MUST implement a gRPC `FunctionRunnerService` server. A Function
+SHOULD implement the latest available version of this service - e.g. `v1beta1`.
+The authoritative definition of this service can be found at the following URL.
+
+https://github.com/crossplane/crossplane/tree/master/apis/apiextensions/fn/proto
+
+A Function MUST copy the tag field from a RunFunctionRequest's RequestMeta
+message to the ResponseMeta tag field of the corresponding RunFunctionResponse.
+A Function MUST NOT modify the tag field. A Function MUST NOT inspect or
+otherwise depend on the tag's value
+
+A Function SHOULD specify the ttl field of each RunFunctionResponse. The value
+of the ttl field MUST reflect how long Crossplane may use a cached version of
+the response for an identical RunFunctionRequest.
+
+A Function MUST pass through any desired state for which it does not have an
+opinion from the RunFunctionRequest desired field to the corresponding
+RunFunctionResponse desired field. A Function MAY mutate any desired state for
+which it has an opinion, including adding to, updating, or deleting from the
+desired state. For example a Function may:
+
+* Add a new composed resource to the desired state.
+* Add a new field to an existing composed resource in the desired state.
+* Remove a composed resource from the desired state.
+* etc...
+
+A Function MUST do so _intentionally_. Put otherwise, remember that Functions
+are run in a pipeline. A Function MUST NOT accidentally forget to propagate the
+desired state passed to it by previous Functions.
+
+A Function MUST NOT specify desired composite resource spec or metadata fields.
+A Function MAY only specify desired composite resource status fields.
+
+A Function MUST NOT specify desired composed resource status fields. A Function
+MAY only specify non-status desired composed resource fields, for example fields
+under the top-level metadata, spec, or data fields.
+
+A Function SHOULD NOT specify a metadata.name for desired composed resources.
+Crossplane will generate an appropriate name. A Function SHOULD specify the
+crossplane.io/external-name annotation for its desired composed resources in
+order to influence the external name of those resources.
+
+A Function SHOULD avoid conflicting with existing observed or desired composed
+resources, for example by attempting to add an entry to the desired composed
+resources array with a name that already exists in either that array or the
+observed composed resources array.
+
+A Function MUST only return a Fatal result if it intends to terminate the entire
+Function pipeline, including preventing any Functions that would normally run
+after it from running.
+
+A Function MUST NOT return a Warning or Normal result every time it is called. A
+Function SHOULD return Warning or Normal results to indicate transitions - e.g.
+when entering a warning or normal state.
+
+## Configuration
+
+A Function MUST support the following command-line flags:
```

Comment:
Maybe we could add something about debug logging, this way we could even think of making that configurable at Function level.